In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/raw/DataCoSupplyChainDataset.csv", encoding="latin1")

In [3]:
df. head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [4]:
df.shape

(180519, 53)

In [5]:
df.columns.tolist()

['Type',
 'Days for shipping (real)',
 'Days for shipment (scheduled)',
 'Benefit per order',
 'Sales per customer',
 'Delivery Status',
 'Late_delivery_risk',
 'Category Id',
 'Category Name',
 'Customer City',
 'Customer Country',
 'Customer Email',
 'Customer Fname',
 'Customer Id',
 'Customer Lname',
 'Customer Password',
 'Customer Segment',
 'Customer State',
 'Customer Street',
 'Customer Zipcode',
 'Department Id',
 'Department Name',
 'Latitude',
 'Longitude',
 'Market',
 'Order City',
 'Order Country',
 'Order Customer Id',
 'order date (DateOrders)',
 'Order Id',
 'Order Item Cardprod Id',
 'Order Item Discount',
 'Order Item Discount Rate',
 'Order Item Id',
 'Order Item Product Price',
 'Order Item Profit Ratio',
 'Order Item Quantity',
 'Sales',
 'Order Item Total',
 'Order Profit Per Order',
 'Order Region',
 'Order State',
 'Order Status',
 'Order Zipcode',
 'Product Card Id',
 'Product Category Id',
 'Product Description',
 'Product Image',
 'Product Name',
 'Product P

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  str    
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  str    
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  str    
 9   Customer City                  180519 non-null  str    
 10  Customer Country               180519 non-null  str    
 11  Customer Email                 180519 non-null  str    
 12  Customer Fname                 180519 non

In [7]:
df.isnull().sum()

Type                                  0
Days for shipping (real)              0
Days for shipment (scheduled)         0
Benefit per order                     0
Sales per customer                    0
Delivery Status                       0
Late_delivery_risk                    0
Category Id                           0
Category Name                         0
Customer City                         0
Customer Country                      0
Customer Email                        0
Customer Fname                        0
Customer Id                           0
Customer Lname                        8
Customer Password                     0
Customer Segment                      0
Customer State                        0
Customer Street                       0
Customer Zipcode                      3
Department Id                         0
Department Name                       0
Latitude                              0
Longitude                             0
Market                                0


In [8]:
df["Delivery Status"].value_counts()

Delivery Status
Late delivery        98977
Advance shipping     41592
Shipping on time     32196
Shipping canceled     7754
Name: count, dtype: int64

In [9]:
df["Late_delivery_risk"].value_counts()

Late_delivery_risk
1    98977
0    81542
Name: count, dtype: int64

In [10]:
round(df["Late_delivery_risk"].mean() * 100, 2)

np.float64(54.83)

In [11]:
df.groupby("Shipping Mode")["Late_delivery_risk"].mean()

Shipping Mode
First Class       0.953225
Same Day          0.457430
Second Class      0.766328
Standard Class    0.380717
Name: Late_delivery_risk, dtype: float64

In [12]:
df["Shipping Mode"].value_counts()

Shipping Mode
Standard Class    107752
Second Class       35216
First Class        27814
Same Day            9737
Name: count, dtype: int64

In [13]:
df.groupby("Shipping Mode")["Late_delivery_risk"].mean().sort_values(ascending=False)

Shipping Mode
First Class       0.953225
Second Class      0.766328
Same Day          0.457430
Standard Class    0.380717
Name: Late_delivery_risk, dtype: float64

In [14]:
df.groupby("Market")["Late_delivery_risk"].mean().sort_values(ascending=False)

Market
Europe          0.552078
Pacific Asia    0.550460
USCA            0.548006
Africa          0.545893
LATAM           0.543552
Name: Late_delivery_risk, dtype: float64

In [15]:
df.groupby("Department Name")["Late_delivery_risk"].mean().sort_values(ascending=False)

Department Name
Pet Shop              0.589431
Book Shop             0.565432
Health and Beauty     0.558011
Fitness               0.555466
Outdoors              0.554925
Technology            0.550171
Golf                  0.547803
Fan Shop              0.547748
Apparel               0.547471
Footwear              0.547263
Discs Shop            0.544423
Name: Late_delivery_risk, dtype: float64

In [16]:
df.groupby("Type")["Late_delivery_risk"].mean()

Type
CASH        0.566323
DEBIT       0.572177
PAYMENT     0.575291
TRANSFER    0.485436
Name: Late_delivery_risk, dtype: float64

In [17]:
df[["Days for shipping (real)","Days for shipment (scheduled)"]].describe()

,Days for shipping (real),Days for shipment (scheduled)
count,180519.000000,180519.000000
mean,3.497654,2.931847
std,1.623722,1.374449
min,0.000000,0.000000
25%,2.000000,2.000000
50%,3.000000,4.000000
75%,5.000000,4.000000
max,6.000000,4.000000


In [18]:
df["Shipping_delay_days"]=df["Days for shipping (real)"]-df["Days for shipment (scheduled)"]

In [19]:
df["Shipping_delay_days"].describe()

count    180519.000000
mean          0.565807
std           1.490966
min          -2.000000
25%           0.000000
50%           1.000000
75%           1.000000
max           4.000000
Name: Shipping_delay_days, dtype: float64

In [20]:
df.shape

(180519, 54)

In [21]:
[col for col in df.columns if "delay" in col.lower()]

['Shipping_delay_days']

In [22]:
df["Shipping_delay_days"].value_counts().sort_index()


Shipping_delay_days
-2    21666
-1    21700
 0    33753
 1    60647
 2    28718
 3     7052
 4     6983
Name: count, dtype: int64

# Key Findings

## Delivery Performance

- 54.83% of records were flagged with late-delivery risk.
- 57.28% of records had a positive shipping delay and were classified as Delayed.
- Only 18.70% of records were delivered exactly on time.
- Most delays were between 1 and 2 days.

## Shipping Modes

- First Class had the highest late delivery risk (95.3%).
- Second Class had the second highest late delivery risk (76.6%).
- Standard Class showed the best performance among all shipping methods.

## Market Analysis

- Late delivery risk was consistent across all markets (~54-55%).
- No significant regional delivery problem was identified.

## Department Analysis

- Late delivery risk was similar across product departments.
- Product category was not a major driver of delays.

## Root Cause Hypothesis

- Actual shipping time exceeds scheduled shipping time by an average of 0.57 days.
- Delivery promises appear more aggressive than operational capability.

In [23]:
df_clean = df.copy()

In [24]:
columns_to_remove = [
    "Customer Email",
    "Customer Password",
    "Customer Street",
    "Customer Fname",
    "Customer Lname"
]

df_clean = df_clean.drop(columns=columns_to_remove)

In [25]:
df_clean.columns = (
    df_clean.columns
    .str.lower()
    .str.strip()
    .str.replace(" ", "_")
    .str.replace("(", "")
    .str.replace(")", "")
)

In [26]:
df_clean.columns.tolist()[:10]

['type',
 'days_for_shipping_real',
 'days_for_shipment_scheduled',
 'benefit_per_order',
 'sales_per_customer',
 'delivery_status',
 'late_delivery_risk',
 'category_id',
 'category_name',
 'customer_city']

In [27]:
def delivery_category(delay):
    if delay < 0:
        return "Early"
    elif delay == 0:
        return "On Time"
    else:
        return "Delayed"

In [28]:
df_clean["delivery_performance"] = df_clean["shipping_delay_days"].apply(delivery_category)

In [29]:
df_clean["delivery_performance"].value_counts()

delivery_performance
Delayed    103400
Early       43366
On Time     33753
Name: count, dtype: int64

In [30]:
df_clean["profit_margin"] = (
    (
        df_clean["benefit_per_order"]
        / df_clean["sales_per_customer"]
    ) * 100
).round(6)

In [31]:
df_clean["profit_margin"].describe()

count    180519.000000
mean         12.044936
std          46.665982
min        -275.081264
25%           7.496664
50%          27.001340
75%          36.299814
max          50.061051
Name: profit_margin, dtype: float64

In [32]:
df_clean[df_clean["profit_margin"] < 0]["profit_margin"].describe()

count    33784.000000
mean       -62.464734
std         62.707423
min       -275.081264
25%        -79.997948
50%        -38.302421
75%        -16.701249
max         -1.091486
Name: profit_margin, dtype: float64

In [33]:
df_clean[df_clean["profit_margin"] < -100][
    ["benefit_per_order", "sales_per_customer", "profit_margin"]
].head()

,benefit_per_order,sales_per_customer,profit_margin
34,-425.579987,321.200012,-132.496877
49,-122.730003,79.180000,-155.001267
75,-285.070007,178.169998,-159.998883
125,-783.669983,474.950012,-165.000519
188,-284.529999,167.369995,-170.000602


In [34]:
def profitability_category(margin):
    if margin < 0:
        return "Loss"
    elif margin < 20:
        return "Low Margin"
    else:
        return "Healthy Margin"

In [35]:
df_clean["profitability_category"] = df_clean["profit_margin"].apply(profitability_category)

In [36]:
df_clean["profitability_category"].value_counts()

profitability_category
Healthy Margin    107456
Low Margin         39279
Loss               33784
Name: count, dtype: int64

In [37]:
pd.crosstab(
    df_clean["delivery_performance"],
    df_clean["profitability_category"]
)

profitability_category,Healthy Margin,Loss,Low Margin
delivery_performance,,,
Delayed,61441,19421,22538
Early,25810,8095,9461
On Time,20205,6268,7280


In [38]:
pd.crosstab(
    df_clean["delivery_performance"],
    df_clean["profitability_category"],
    normalize="index"
) * 100

profitability_category,Healthy Margin,Loss,Low Margin
delivery_performance,,,
Delayed,59.420696,18.782398,21.796905
Early,59.516672,18.666697,21.816631
On Time,59.861346,18.570201,21.568453


In [39]:
df_clean.groupby("category_name")["profit_margin"].mean().sort_values()

category_name
Men's Clothing           5.279273
Strength Training        5.926431
As Seen on  TV!          6.472824
Books                    7.868961
Video Games              9.152930
Hunting & Shooting       9.263778
Pet Supplies             9.394630
Health and Beauty        9.540617
DVDs                     9.623247
Golf Apparel             9.975241
Kids' Golf Clubs        10.534263
Basketball              10.781205
Shop By Sport           11.205213
Boxing & MMA            11.247725
Camping & Hiking        11.572148
Water Sports            11.628541
Computers               11.687347
Golf Balls              11.799509
Girls' Apparel          11.829060
Cardio Equipment        11.864760
Men's Footwear          11.991375
Sporting Goods          12.087972
Golf Shoes              12.113494
Fishing                 12.127139
Women's Apparel         12.200665
Indoor/Outdoor Games    12.373216
Cleats                  12.442688
Electronics             12.444628
Crafts                  12.517776


In [41]:
category_performance = (
    df_clean.groupby("category_name")
    .agg(
        total_sales=("sales", "sum"),
        total_profit=("order_profit_per_order", "sum"),
        average_profit_margin=("profit_margin", "mean"),
        unique_orders=("order_id", "nunique")
    )
    .sort_values("total_profit", ascending=False)
)

category_performance

,total_sales,total_profit,average_profit_margin,unique_orders
category_name,,,,
Fishing,6.929654e+06,756220.767190,12.127139,15164
Cleats,4.431943e+06,494636.919791,12.442688,20386
Camping & Hiking,4.118426e+06,427455.568106,11.572148,12299
Cardio Equipment,3.694843e+06,383011.098485,11.864760,11355
Women's Apparel,3.147800e+06,350421.029567,12.200665,17869
Water Sports,3.113845e+06,325146.960038,11.628541,13758
Indoor/Outdoor Games,2.888994e+06,318451.430554,12.373216,16623
Men's Footwear,2.891758e+06,311902.820214,11.991375,18783
Shop By Sport,1.309522e+06,129813.960315,11.205213,10136


In [42]:
df_clean.columns.tolist()

['type',
 'days_for_shipping_real',
 'days_for_shipment_scheduled',
 'benefit_per_order',
 'sales_per_customer',
 'delivery_status',
 'late_delivery_risk',
 'category_id',
 'category_name',
 'customer_city',
 'customer_country',
 'customer_id',
 'customer_segment',
 'customer_state',
 'customer_zipcode',
 'department_id',
 'department_name',
 'latitude',
 'longitude',
 'market',
 'order_city',
 'order_country',
 'order_customer_id',
 'order_date_dateorders',
 'order_id',
 'order_item_cardprod_id',
 'order_item_discount',
 'order_item_discount_rate',
 'order_item_id',
 'order_item_product_price',
 'order_item_profit_ratio',
 'order_item_quantity',
 'sales',
 'order_item_total',
 'order_profit_per_order',
 'order_region',
 'order_state',
 'order_status',
 'order_zipcode',
 'product_card_id',
 'product_category_id',
 'product_description',
 'product_image',
 'product_name',
 'product_price',
 'product_status',
 'shipping_date_dateorders',
 'shipping_mode',
 'shipping_delay_days',
 'delive

In [43]:
columns_to_remove = [
    "category_id",
    "department_id",
    "order_customer_id",
    "order_item_cardprod_id",
    "order_item_id",
    "product_card_id",
    "product_category_id",
    "customer_city",
    "customer_state",
    "customer_zipcode",
    "latitude",
    "longitude",
    "product_description",
    "product_image",
    "order_item_product_price",
    "product_status"
]
df_clean = df_clean.drop(columns= columns_to_remove)

In [44]:
df_clean.shape

(180519, 36)

In [45]:
df_clean.isnull().sum().sort_values(ascending= False).head(10)

order_zipcode             155679
days_for_shipping_real         0
order_item_quantity            0
sales                          0
order_item_total               0
order_profit_per_order         0
order_region                   0
order_state                    0
order_status                   0
type                           0
dtype: int64

In [46]:
df_clean.columns.tolist()

['type',
 'days_for_shipping_real',
 'days_for_shipment_scheduled',
 'benefit_per_order',
 'sales_per_customer',
 'delivery_status',
 'late_delivery_risk',
 'category_name',
 'customer_country',
 'customer_id',
 'customer_segment',
 'department_name',
 'market',
 'order_city',
 'order_country',
 'order_date_dateorders',
 'order_id',
 'order_item_discount',
 'order_item_discount_rate',
 'order_item_profit_ratio',
 'order_item_quantity',
 'sales',
 'order_item_total',
 'order_profit_per_order',
 'order_region',
 'order_state',
 'order_status',
 'order_zipcode',
 'product_name',
 'product_price',
 'shipping_date_dateorders',
 'shipping_mode',
 'shipping_delay_days',
 'delivery_performance',
 'profit_margin',
 'profitability_category']

In [47]:
df_clean=df_clean.drop(columns=["order_zipcode"])

In [48]:
df_clean.columns.tolist()

['type',
 'days_for_shipping_real',
 'days_for_shipment_scheduled',
 'benefit_per_order',
 'sales_per_customer',
 'delivery_status',
 'late_delivery_risk',
 'category_name',
 'customer_country',
 'customer_id',
 'customer_segment',
 'department_name',
 'market',
 'order_city',
 'order_country',
 'order_date_dateorders',
 'order_id',
 'order_item_discount',
 'order_item_discount_rate',
 'order_item_profit_ratio',
 'order_item_quantity',
 'sales',
 'order_item_total',
 'order_profit_per_order',
 'order_region',
 'order_state',
 'order_status',
 'product_name',
 'product_price',
 'shipping_date_dateorders',
 'shipping_mode',
 'shipping_delay_days',
 'delivery_performance',
 'profit_margin',
 'profitability_category']

In [49]:
df_clean.isnull().sum().sort_values(ascending=False).head(10)

type                        0
order_status                0
order_item_quantity         0
sales                       0
order_item_total            0
order_profit_per_order      0
order_region                0
order_state                 0
product_name                0
order_item_discount_rate    0
dtype: int64

In [50]:
df_clean.shape

(180519, 35)

In [51]:
df_clean.duplicated().sum()

np.int64(0)

In [52]:
df_clean.dtypes[["order_date_dateorders", "shipping_date_dateorders"]]

order_date_dateorders       str
shipping_date_dateorders    str
dtype: object

In [53]:
df_clean["order_date_dateorders"] = pd.to_datetime(
    df_clean["order_date_dateorders"]
)

df_clean["shipping_date_dateorders"] = pd.to_datetime(
    df_clean["shipping_date_dateorders"]
)

In [54]:
df_clean.dtypes[['order_date_dateorders','shipping_date_dateorders']]

order_date_dateorders       datetime64[us]
shipping_date_dateorders    datetime64[us]
dtype: object

In [55]:
df_clean.to_csv(
    "../data/processed/supply_chain_cleaned.csv",
    index=False
)

### Regional Delivery Analysis

- Delivery risk was relatively consistent across regions.
- The highest delay risk was Central Africa (57.96%).
- The lowest delay risk was Canada (48.80%).
- Regional differences were much smaller compared to shipping mode differences.
- Geography does not appear to be the main driver of delivery delays.

### Category Profitability Analysis

- Fishing generated the highest sales ($6.93M) and profit ($756K).
- Cleats generated $4.43M in sales and $495K in profit.
- Camping & Hiking generated $4.12M in sales and $427K in profit.
- Revenue and profit are concentrated within a small number of categories.
- Categories with the highest margins are not necessarily the most profitable.
- Fishing outperformed higher-margin categories due to significantly higher sales volume.
- Computers showed unusually high revenue per order despite a low number of orders.

### Low Performing Categories

- Strength Training generated only $332 profit despite $54.9K in sales.
- Men's Clothing showed a low profit margin of 5.28%.
- Books and Video Games generated relatively low profit compared to sales volume.
- Product category performance varies significantly across the business.
- Category selection appears to have a stronger impact on profitability than customer segment or geographic region.

### Discount Impact Analysis

- Discounted orders represented the majority of transactions.
- Low discount orders generated the highest sales ($16.35M) and profit ($1.88M).
- Increasing discounts reduced average profit margin.
- High discounts did not generate proportional profit growth.
- Low discount strategies appear to provide the best balance between sales volume and profitability.

### Customer Segment and Category Analysis

- Consumer customers are the largest drivers of profitable categories.
- Consumer + Fishing generated the highest profit combination ($393K).
- Fishing performed strongly across all customer segments, indicating broad market demand.
- Corporate customers showed unexpected strength in Women's Apparel, generating nearly $1M in sales and $109K profit.
- Customer segmentation reveals purchasing patterns that are hidden when analysing categories alone.

### Country-Level Profitability Analysis

- USA generated the highest overall sales ($4.88M) and profit ($540K).
- France and Mexico showed strong profitability with above-average margins.
- Germany generated significant revenue but had a lower average margin (11.21%) compared with the company average (12.04%).
- Country-level profitability differences appear to be driven mainly by sales volume rather than large margin differences.
- Further investigation is required into product mix and customer behaviour by market.

### Country and Category Analysis

- Fishing is the strongest global category, generating significant profit across multiple countries.
- USA profitability is driven by a diverse portfolio including Fishing, Camping & Hiking, Cleats, and Women's Apparel.
- Germany shows strong performance in Fishing products despite having a lower overall country margin.
- Women's Apparel demonstrates strong profitability in France, Mexico, and USA markets.
- Product mix appears to be a major driver of country-level profitability.

### Country + Category + Delivery Analysis

- High-value categories are frequently affected by delivery delays.
- USA Fishing orders represent the largest delayed profit exposure.
- Germany Fishing shows strong profitability but significant delayed orders.
- Delay duration is consistent across markets (~1.6 days), suggesting operational process issues rather than isolated geographic problems.
- Supply chain improvements should prioritize high-profit category and market combinations.